internal/scheduler/metrics/metrics.go

```diff
@@ -477,72 +459,91 @@ func errorTypeAndMessageFromError(ctx *armadacontext.Context, err *armadaevents.
+	if c, err := m.jobs.GetMetricWithLabelValues(labels[1:]...); err != nil { // we don't need priorState label here
 		return err
 	} else {
 		c.Add(1)
 	}
-
-	// Total resource requests of jobs.
-	requests := job.GetResourceRequirements().Requests
-	for _, resourceName := range m.config.TrackedResourceNames {
-		labels[i] = string(resourceName)
-		if c, err := vec.GetMetricWithLabelValues(labels...); err != nil {
-			return err
-		} else {
-			q := requests[resourceName]
-			v := float64(q.MilliValue()) / 1000
-			c.Add(v)
-		}
-	}
-
-	return nil
-}
-
-// updateResourceSecondsCounterVec is a helper method to increment vector counters by the number of seconds per resource a jobs has consumed in a given state.
-func (m *Metrics) updateResourceSecondsCounterVec(vec *prometheus.CounterVec, labels []string, job *jobdb.Job, stateTime, priorStateTime *time.Time) error {
-	if stateTime == nil || priorStateTime == nil {
-		return nil
-	}
-	i := len(labels)
-	// Number of jobs.
-	labels = append(labels, jobsResourceLabel)
-	stateDuration := stateDuration(priorStateTime, stateTime).Seconds()
-	if c, err := vec.GetMetricWithLabelValues(labels...); err != nil {
+	// update jobs-seconds counter
+	if c, err := m.jobsSeconds.GetMetricWithLabelValues(labels...); err != nil {
 		return err
 	} else {
-		c.Add(stateDuration)
+		c.Add(stateDuration.Seconds())
 	}
+
 	requests := job.GetResourceRequirements().Requests
-	for _, resourceName := range m.config.TrackedResourceNames {
-		labels[i] = string(resourceName)
-		if c, err := vec.GetMetricWithLabelValues(labels...); err != nil {
+	for _, resource := range m.config.TrackedResourceNames {
+		if r, ok := m.config.ResourceRenaming[resource]; ok {
+			resource = v1.ResourceName(r)
+		}
+		if !metricNameValidationRegex.MatchString(resource.String()) {
+			logrus.Warnf("Resource name is not valid for a metric name: %s", resource)
+			continue
+		}
+		metric, metricSeconds := m.getOrCreateMetrics(resource)
+		if metric == nil || metricSeconds == nil {
+			continue
+		}
+		c, err := metric.GetMetricWithLabelValues(labels[1:]...) // we don't need priorState label here
+		if err != nil {
 			return err
-		} else {
-			q := requests[resourceName]
-			v := float64(q.MilliValue()) / 1000
-			c.Add(v * stateDuration)
 		}
+		cSeconds, err := metricSeconds.GetMetricWithLabelValues(labels...)
+		if err != nil {
+			return err
+		}
+		q := requests[resource]
+		v := float64(q.MilliValue()) / 1000
+		c.Add(v)
+		cSeconds.Add(v * stateDuration.Seconds())
 	}
 	return nil
 }
 
-func stateDuration(start *time.Time, end *time.Time) time.Duration {
-	if start != nil && end != nil {
-		return end.Sub(*start)
+// getOrCreateMetrics returns the counter and counterSeconds Vec for the given resource name.
+// If the counter and counterSeconds Vecs do not exist, they are created and stored in the resourceCounters map.
+func (m *Metrics) getOrCreateMetrics(resource v1.ResourceName) (*prometheus.CounterVec, *prometheus.CounterVec) {
```

Comment:
```suggestion
func (m *Metrics) counterVectorsFromResource(resource v1.ResourceName) (*prometheus.CounterVec, *prometheus.CounterVec) {
```

Reply:
You can do this also for jobs to avoid creating it eagerly in New(). A bit cleaner.